src/java.base/share/native/libjli/java.c

```diff
@@ -534,6 +582,329 @@ invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
+ * @param[in] argc The number of command line arguments
+ * @param[in] argv The array of command line arguments
+ * @param[out] error A pointer to an integer for the error code
+ * @return A pointer to the value of the command line option if found, NULL otherwise
+ */
+static const char *
+getCommandLineOptionValue(const char *optionName, int argc, char **argv, int *error)
+{
+    const char *value = NULL;
+    int i = 0;
+    size_t optionNameLength = strlen(optionName);
+    jboolean optionNameFound = JNI_FALSE;
+    *error = 0;
+    for (; i < argc; i++) {
+        const char *arg = argv[i];
+        if (0 == strncmp(arg, optionName, optionNameLength)) {
+            const char *equals = arg + optionNameLength;
+            if (('=' == *equals) || ('\0' == *equals)) {
+                optionNameFound = JNI_TRUE;
+                value = equals;
+                if ('=' == *value) {
+                    value = value + 1;
+                }
+                if ('\0' == *value) {
+                    value = NULL;
+                }
+                break;
+            }
+        }
+    }
+    if (!optionNameFound) {
+        *error = OPTION_NAME_NOT_FOUND_ERROR;
+    }
+    return value;
+}
+
+/**
+ * Get the checkpoint directory from command line options.
+ * @param[in] argc The number of command line arguments
+ * @param[in] argv The array of command line arguments
+ * @param[in/out] error A pointer to an integer for the error code
+ * @return A pointer to the checkpoint directory if found, NULL otherwise
+ */
+static const char *
+getCheckpointDirectory(int argc, char **argv, int *error)
+{
+    const char *checkpointDirectory = NULL;
+    const char *checkpointDirectoryPropertyValue = getCommandLineOptionValue(CRAC_RESTORE_FROM_OPTION_NAME, argc, argv, error);
+    if (0 == *error) {
+        if (NULL != checkpointDirectoryPropertyValue) {
+            checkpointDirectory = checkpointDirectoryPropertyValue;
+        } else {
+            JLI_ReportErrorMessage(OPTION_VALUE_NOT_FOUND_ERROR_MESSAGE, CRAC_RESTORE_FROM_OPTION_NAME);
+            *error = OPTION_VALUE_NOT_FOUND_ERROR;
+        }
+    }
+    return checkpointDirectory;
+}
+
+/**
+ * Check if a command line option is found with an error condition.
+ * @param[in] error The error code of the result of obtaining the command line option
+ * @return true if the command line option is found and an error occurred, false otherwise
+ */
+static jboolean
+isCommandLineOptionFoundWithError(int error)
+{
+    return (OPTION_NAME_NOT_FOUND_ERROR != error) && (0 != error);
+}
+
+/**
+ * Get the log level specified in the command line arguments.
+ * @param[in] argc The number of command line arguments
+ * @param[in] argv The array of command line arguments
+ * @param[in/out] error A pointer to an integer for the error code
+ * @return A pointer to the log level string if successful, NULL otherwise
+ */
+static const char *
+getLogLevel(int argc, char **argv, int *error)
+{
```

Comment:
Suggest this should return `int`, avoiding the need to call `atoi()` other places.
```c
static int
getLogLevel(int argc, char **argv, int *error)
{
    int logLevelValue = 2; /* default */
    const char *logLevelPropertyValue = getCommandLineOptionValue(LOG_LEVEL_OPTION_NAME, argc, argv, error);
    if (!isCommandLineOptionFoundWithError(*error)) {

        /* [elided] */

        if ((0 <= logLevelValue) && (logLevelValue <= 4)) {
            goto done;
        }
    }
setLogLevelOptionValueNotValidError:
    JLI_ReportErrorMessage(
            "The value of the command line option %s is not valid, value=%s.",
            LOG_LEVEL_OPTION_NAME,
            (NULL != logLevelPropertyValue) ? logLevelPropertyValue : "");
    *error = OPTION_VALUE_NOT_VALID_ERROR;
done:
    return logLevelValue;
}
```